File: 0039-combination-sum/0039-combination-sum.java
class Solution {
    public List<List<Integer>> combinationSum(int[] candidates, int target) {
        List<List<Integer>> result = new ArrayList<>();

        dfs(result, candidates, target, 0, new ArrayList<Integer>());
        return result;
    }

    private void dfs(List<List<Integer>> result, int[] candidates, 
                    int target, int prev, List<Integer> cur) {
        if (target == 0) {
            result.add(new ArrayList<>(cur));
            return;
        }

        for (int i = prev; i < candidates.length; i++) {
            if (candidates[i] <= target) {
                cur.add(candidates[i]);
                dfs(result, candidates, target - candidates[i], i, cur);
                cur.remove(cur.size() - 1);
            }
        }
    }
}

File: 0039-combination-sum/0039-combination-sum.cpp
class Solution {
private:
    vector<vector<int>> combinations;
    vector<int> cur;

public:
    vector<vector<int>> combinationSum(vector<int>& candidates, int target) {
        backtrack(candidates, target, 0);
        return combinations;
    }

private:
    void backtrack(vector<int>& candidates, int target, int index) {
        if (target <= 0) {
            if (target == 0) {
                combinations.push_back(cur);
            }
            return;
        }

        for (int i = index; i < candidates.size(); i++) {
            cur.push_back(candidates[i]);
            backtrack(candidates, target - candidates[i], i);
            cur.pop_back();
        }
    }
};